exec/exec.go

```diff
@@ -10,37 +11,82 @@ import (
 
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/pkg/rand"
 )
 
 var ErrWaitPIDTimeout = fmt.Errorf("Timed out waiting for PID to complete")
 
+type CmdOpts struct {
+	timeout time.Duration
+}
+
+var DefaultCmdOpts = CmdOpts{
+	timeout: time.Duration(0),
+}
+
 // RunCommandExt is a convenience function to run/log a command and return/log stderr in an error upon
-// failure. Chops off a trailing newline (if present)
-func RunCommandExt(cmd *exec.Cmd) (string, error) {
-	cmdStr := strings.Join(cmd.Args, " ")
-	log.Info(cmdStr)
-	outBytes, err := cmd.Output()
+// failure.
+func RunCommandExt(cmd *exec.Cmd, opts CmdOpts) (string, error) {
+
+	logCtx := log.WithFields(log.Fields{"execID": rand.RandString(5)})
+	// log in a way we can copy-and-paste into a terminal
+	args := strings.Join(cmd.Args, " ")
+	logCtx.WithFields(log.Fields{"dir": cmd.Dir}).Info(args)
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	start := time.Now()
+	err := cmd.Start()
 	if err != nil {
-		exErr, ok := err.(*exec.ExitError)
```

Comment:
@alexec why did you drop special handling for `exec.ExitError` ? Won't we loose stderr output because of this?

Reply:
as discussed, this cannot panic